From 4c583c44a80d75acb86f818c57e0eeecac25a462 Mon Sep 17 00:00:00 2001
From: Krisztian Litkey <krisztian.litkey@intel.com>
Date: Tue, 8 Dec 2015 16:02:56 +0200
Subject: [PATCH 1/1] nspawn: don't check binary/OS-tree before all mounts are
 in place.

Do not check if the binary to run is in place or if the given
directory looks like an OS tree hierarchy *before* we have put
all the requested bind- and overlay-mounts in place. It is
possible that some of the mounts are responsible for pulling in
the necessary bits and pieces for a successful check.
---
 src/nspawn/nspawn.c | 36 +++++++++++++++++-------------------
 1 file changed, 17 insertions(+), 19 deletions(-)

diff --git a/src/nspawn/nspawn.c b/src/nspawn/nspawn.c
index 3428109..68bcb9f 100644
--- a/src/nspawn/nspawn.c
+++ b/src/nspawn/nspawn.c
@@ -4395,6 +4395,23 @@ static int outer_child(
         if (r < 0)
                 return r;
 
+        /* Now that everything is mounted, check if we have the necessary binary/directory tree to continue. */
+        if (arg_boot) {
+                if (path_is_os_tree(directory) <= 0) {
+                        log_error("Directory %s doesn't look like an OS root directory (os-release file is missing). Refusing.", directory);
+                        return -EINVAL;
+                } else {
+                        const char *p;
+
+                        p = strjoina("/", argc > optind && path_is_absolute(argv[optind]) ? argv[optind] : "/usr/bin/");
+
+                        if (access(p, F_OK) < 0) {
+                                log_error("Directory %s lacks the binary to execute or doesn't look like a binary tree. Refusing.", directory);
+                                return -EINVAL;
+                        }
+                }
+        }
+
         r = mount_cgroup(directory);
         if (r < 0)
                 return r;
@@ -4602,25 +4619,6 @@ int main(int argc, char *argv[]) {
                                 }
                         }
                 }
-
-                if (arg_boot) {
-                        if (path_is_os_tree(arg_directory) <= 0) {
-                                log_error("Directory %s doesn't look like an OS root directory (os-release file is missing). Refusing.", arg_directory);
-                                r = -EINVAL;
-                                goto finish;
-                        }
-                } else {
-                        const char *p;
-
-                        p = strjoina(arg_directory,
-                                       argc > optind && path_is_absolute(argv[optind]) ? argv[optind] : "/usr/bin/");
-                        if (access(p, F_OK) < 0) {
-                                log_error("Directory %s lacks the binary to execute or doesn't look like a binary tree. Refusing.", arg_directory);
-                                r = -EINVAL;
-                                goto finish;
-                        }
-                }
-
         } else {
                 char template[] = "/tmp/nspawn-root-XXXXXX";
 
-- 
2.5.0

